e2e/pkg/installation_builder.go

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+//
+
+//+build e2e
+
+package pkg
+
+import "github.com/mattermost/mattermost-cloud/model"
+
+// InstallationBuilder is a helper to create CreateInstallationRequest.
+type InstallationBuilder struct {
+	request *model.CreateInstallationRequest
+}
+
+// NewInstallationBuilderWithDefaults sets up new InstallationBuilder with resonable defaults.
+func NewInstallationBuilderWithDefaults() *InstallationBuilder {
```

Comment:
Couldn't this just be a function `NewInstallationWithDefaults()` that returns a `&model.CreateInstallationRequest` and then you don't need all the helper methods or the `InstallationBuilder` type? It's not clear to me what the value of this type is, especially since it isn't a real factory pattern, it's just a type that wraps (without embedding) the `CreateInstallationRequest` so you need a new `InstallationBuilder` for each `CreateInstallationRequest`, which seems to be roughly the same as if you had a function that simply returned `CreateInstallationRequest`s with predefined defaults, but with more indirection.

There's also a strong possibility that I'm not understanding something

Reply:
Yeah, so this function does not contain all necessary parameters for creating installation - mainly DNS - and others can be parametrized by the test itself.

The idea is that `NewCreateInstallationRequestWithDefaults()` gives you some base for which you can then override fields that you need in a more elegant way (at least in my opinion 😄 ) than just directly changing struct values. This allows you to add your customization by chaining methods. There is an example in one of the test steps:
```go
installationBuilder := pkg.NewInstallationBuilderWithDefaults().
	DNS(pkg.GetDNS(w.env)).
	DB(w.Params.DBType).
	FileStore(w.Params.FileStoreType)
```

> It's not clear to me what the value of this type is, especially since it isn't a real factory pattern

Not sure if I understand what do you mean here. In my understanding factory pattern is something different and it does not aim to be one - it is a simple builder. That said the interpretations of different patterns nowadays vary wildely especially across different languages so we just might have a different understanding.

Reply:
Seeing the usage helps me understand the pattern that you were going for. I thought it was intended to be a Factory pattern (I've seen Factories called Builders before, so that's the source of the confusion I think), but like you said -- the Factory pattern is something different. Thanks for the clarification.

Is there a reason we shouldn't add these setters to the `CreateInstallationRequest` method set instead of creating a new type only for testing? I don't think it hurts to have them there, and then you could do (assuming the first method `NewInstallationBuilderWithDefaults` was rewritten to be a function that returns a `*CreateInstallationRequest`) 

```go
installationrequest := pkg.NewCreateInstallationRequestWithDefaults().
    DNS(pkg.GetDNS).
    DB(w.Params.DBType).
    FileStore(w.Params.FileStoreType)
 ```

and then in subsequent operations you can access `installationrequest` without indirection, instead of accessing it as `installationBuidler.request` every time. Or maybe I'm still misunderstanding something.

Anyway don't worry about it too much, the usage example helps me understand why it's done this way and if the above suggestion doesn't strike you as a great idea, it's really not that important.

Reply:
Oh, I see you mean to define it directly on the type. Yeah, that would be possible for sure and would serve the purpose well. The only downside I see would be that methods need to be declared in the same package as their receivers therefore we would need to put it into `model` whereas it would be used just in tests so that is why I went with a different type. 

0/5 on whatever way we want to approach it.

Reply:
>The only downside I see would be that methods need to be declared in the same package as their receivers therefore we would need to put it into model whereas it would be used just in tests so that is why I went with a different type.

Yeah, when I made my original comment I actually thought those methods might already exist on the type in `model`, so I thought this method set was redundant. I don't think it would hurt to add them there instead, but I also don't want to make you do more work on this PR unnecessarily, so I'd say at this point just merge it as-is. 

But thanks for answering my questions, anyway, because it's helped my understanding & if we ever desire to use this pattern in the main codebase, outside of tests, we can always move these methods & modify the test as I suggested in my last comment at that time.

Reply:
> & if we ever desire to use this pattern in the main codebase, outside of tests, we can always move these methods & modify the test

Yep, definitely good with seeing how this works and bringing it into the core code later if we want.